source/fe/mapping_q_cache.cc

```diff
@@ -249,17 +259,454 @@ MappingQCache<dim, spacedim>::initialize(
+          else
+            fe_values->reinit(cell_tria);
+        }
+
+      std::vector<Point<spacedim>> result;
+
+      // Step 2b) read of quadrature points in the relative displacement case
+      // note: we also take this path for non-active or artifical cells so that
+      // these cells are filled with some useful data
+      if (vector_describes_relative_displacement ||
+          (is_active_non_artificial_cell == false))
+        {
+          if (mapping_q_generic != nullptr &&
+              this->get_degree() == mapping_q_generic->get_degree())
+            result =
+              mapping_q_generic->compute_mapping_support_points(cell_tria);
+          else
+            result = fe_values_all.get()->get_quadrature_points();
+
+          // for non-active or artificial cells we are done here and return
+          // the absolute positions, since the provided vector cannot contain
+          // any useful information for these cells
+          if (is_active_non_artificial_cell == false)
+            return result;
+        }
+      else
+        {
+          result.resize(
+            Utilities::pow<unsigned int>(this->get_degree() + 1, dim));
+        }
+
+      // Step 2c) read global vector and adjust points accordingly
+      if ((is_fe_q || is_fe_dgq) && fe.degree == this->get_degree())
+        {
+          // case 1: FE_Q or FE_DGQ with same degree as this class has; this
+          // is the simple case since no interpolation is needed
+          std::vector<types::global_dof_index> dof_indices(
+            fe.n_dofs_per_cell());
+          cell_dofs->get_dof_indices(dof_indices);
+
+          for (unsigned int i = 0; i < dof_indices.size(); ++i)
+            {
+              const auto id = fe.system_to_component_index(i);
+
+              if (is_fe_q)
+                {
+                  // case 1a: FE_Q
+                  if (vector_describes_relative_displacement)
+                    result[id.second][id.first] +=
+                      vector_ghosted(dof_indices[i]);
+                  else
+                    result[id.second][id.first] =
+                      vector_ghosted(dof_indices[i]);
+                }
+              else
+                {
+                  // case 1b: FE_DGQ
+                  if (vector_describes_relative_displacement)
+                    result[lexicographic_to_hierarchic_numbering[id.second]]
+                          [id.first] += vector_ghosted(dof_indices[i]);
+                  else
+                    result[lexicographic_to_hierarchic_numbering[id.second]]
+                          [id.first] = vector_ghosted(dof_indices[i]);
+                }
+            }
+        }
+      else
+        {
+          // case 2: general case; interpolation is needed
+          // note: the following code could be optimized for tensor-product
+          // elements via application of sum factorization as is done on
+          // MatrixFree/FEEvaluation
```

Comment:
Or we could fix `FEValues` to use sum factorization :wink: (but not for this PR).